url49526af437_Maria_G_CS325_tgz-pJ8-assignments_farmer_FarmerImplJ8/src/assignments/frs/chap5/BreakthroughImpl.java

```diff
@@ -0,0 +1,99 @@
+package assignments.frs.chap5;
+
+import assignments.frs.chap5.Breakthrough.PieceType;
+
+/** Implementation stub.
+
+   This source code is from the book 
+     "Flexible, Reliable Software:
+       Using Patterns and Agile Development"
+     published 2010 by CRC Press.
+   Author: 
+     Henrik B Christensen 
+     Computer Science Department
+     Aarhus University
+   
+   This source code is provided WITHOUT ANY WARRANTY either 
+   expressed or implied. You may study, use, modify, and 
+   distribute it for non-commercial purposes. For any 
+   commercial use, see http://www.baerbak.com/
+*/
+public class BreakthroughImpl implements Breakthrough {
+  private PlayerType playerInTurn = PlayerType.WHITE;
+  private Board board = new Board();
+
+	
+  public PieceType getPieceAt( int row, int column ) {
+    return board.get(row,column);
+  }
+  
+  public PlayerType getPlayerInTurn() {
+    return playerInTurn;
+  }
+
+  public PlayerType getWinner() {
+    
+	  for(int col = 0; col < 8; col++){
+		  if( board.get(0, col)  == PieceType.WHITE){
+			  return PlayerType.WHITE;
+		  }
+		  if( board.get(7, col) == PieceType.BLACK){
+			  return PlayerType.BLACK;
+		  }
+	  }
+	  return null;
+  }
+
+  public boolean isMoveValid(int fromRow, int fromColumn,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 61-73 into a separate method.